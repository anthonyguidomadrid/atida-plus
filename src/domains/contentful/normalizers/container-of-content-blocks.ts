import type { Entry } from 'contentful'
import { CmsContentTypes } from '~config/content-types'
import { removeUndefinedPropertiesFromObject } from '~helpers'
import { normalizePromotion } from './promotion'
import { normalizeMarketingTeaser } from './marketing-teaser'
import {
  ContainerOfContentBlocksFragment,
  MarketingTeaserFragment,
  PromotionFragment
} from '~generated-graphql'
import type { ContainerContentBlock } from '~domains/page'

export type ContainerOfContentBlocks = {
  contentType: CmsContentTypes.CONTAINER_OF_CONTENT_BLOCKS
  title?: string
  displayTitle: boolean
  hasSmallMargin: boolean
  blocks?: ContainerContentBlock[]
}

type ContentfulContentBlock = PromotionFragment | MarketingTeaserFragment

export type ContentfulContainerOfContentBlocks = Entry<{
  title?: string
  contentBlocks?: ContentfulContentBlock[]
}>

export const isPromotion = (
  toBeChecked: ContentfulContentBlock
): toBeChecked is PromotionFragment => toBeChecked?.__typename === 'Promotion'

export const isMarketingTeaser = (
  toBeChecked: ContentfulContentBlock
): toBeChecked is MarketingTeaserFragment =>
  toBeChecked?.__typename === 'MarketingTeaser'

const normalizeContainerOfContentBlocksGraphQL = (
  contentBlock: ContainerOfContentBlocksFragment
): ContainerOfContentBlocks | undefined =>
  contentBlock
    ? removeUndefinedPropertiesFromObject({
        contentType: CmsContentTypes.CONTAINER_OF_CONTENT_BLOCKS as const,
        title: contentBlock?.title ?? undefined,
        displayTitle: contentBlock.showTitle ?? false,
        hasSmallMargin: contentBlock.hasMargin ?? false,
        blocks:
          contentBlock?.contentCollection?.items &&
          contentBlock?.contentCollection?.items
            .map(block => {
              if (!block) {
                return null
              }
              if (isMarketingTeaser(block as ContentfulContentBlock))
                return (
                  normalizeMarketingTeaser(block as MarketingTeaserFragment) ??
                  null
                )
              if (isPromotion(block as PromotionFragment))
                return normalizePromotion(block as PromotionFragment) ?? null
              return null
            })
            .filter((block): block is ContainerContentBlock => !!block)
      })
    : undefined

export const normalizeContainerOfContentBlocks = (
  containerOfContentBlocks: ContainerOfContentBlocksFragment
): ContainerOfContentBlocks | undefined => {
  return normalizeContainerOfContentBlocksGraphQL(containerOfContentBlocks)
}
