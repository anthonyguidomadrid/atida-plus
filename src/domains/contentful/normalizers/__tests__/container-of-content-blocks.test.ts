import { normalizeContainerOfContentBlocks } from '../container-of-content-blocks'

const graphQLMock = {
  __typename: 'ContainerOfContentBlocks',
  title: 'Home page marketing teasers',
  showTitle: true,
  contentCollection: {
    items: [
      {
        __typename: 'MarketingTeaser',
        title: 'Boost your immunity PT',
        teaserDescription: 'High quality vitamins for a low price. PT',
        text: '20 % Discount PT',
        color: {
          __typename: 'Color',
          ref: 'secondary-portland-orange'
        },
        image: null,
        bgColor: {
          __typename: 'Color',
          ref: 'category-beauty'
        },
        button: true,
        slug: 'boost-your-immunity-pt'
      },
      {
        __typename: 'MarketingTeaser',
        title: 'Save over 50% with Atida Plus PT',
        description: null,
        text: 'Exclusive offer PT',
        color: {
          __typename: 'Color',
          ref: 'primary-oxford-blue'
        },
        image: {
          __typename: 'Asset',
          url:
            'https://images.ctfassets.net/7g2w796onies/6EOfodeKNE2uKYxJTrpbrA/2b730c8a3d7b66e27a4a78863c7a5156/ac6bb2e17841f7dd4ebd7b1fd77f0185.png'
        },
        bgColor: {
          __typename: 'Color',
          ref: 'category-nutrition-exercise-and-weight-loss'
        },
        button: false,
        slug: 'save-over-50-with-atida-plus'
      }
    ]
  }
} as const

const promotionGraphQLMock = {
  __typename: 'ContainerOfContentBlocks',
  title: 'Home page Promotion Blocks',
  showTitle: false,
  contentCollection: {
    items: [
      {
        __typename: 'Promotion',
        title: 'Avène',
        id: 'Avene has an e with french accent',
        description: {
          __typename: 'PromotionDescription',
          json: {
            data: {},
            content: [
              {
                data: {},
                content: [
                  {
                    data: {},
                    marks: [],
                    value:
                      "We're here to help. Ask about our products or get expert medical advice.",
                    nodeType: 'text'
                  }
                ],
                nodeType: 'paragraph'
              }
            ],
            nodeType: 'document'
          }
        },
        teaserDescription: 'oh, shit...Monday came in',
        isContentWithImage: false,
        labelsCollection: {
          items: [
            {
              __typename: 'Translation',
              key: 'campaign-promo.black-friday.promo.2021',
              value: 'Whoop whoop'
            },
            {
              __typename: 'Translation',
              key: 'campaign-promo.black-friday',
              value: 'Whoop whoop 2'
            }
          ]
        },
        color: {
          __typename: 'Color',
          ref: 'category-beauty'
        },
        image: {
          __typename: 'Asset',
          title: 'Beauty Category Image',
          url:
            '//images.ctfassets.net/7g2w796onies/3Pa3xwIiNIsKGQB2SE2KYy/9045bf905aba7e22e69054b4419c58a9/category-header-sample.png',
          contentType: 'image/png'
        }
      },
      {
        __typename: 'Promotion',
        title: 'Avène',
        id: 'Avene has an e with french accent',
        description: {
          __typename: 'PromotionDescription',
          json: {
            data: {},
            content: [
              {
                data: {},
                content: [
                  {
                    data: {},
                    marks: [],
                    value:
                      "We're here to help. Ask about our products or get expert medical advice.",
                    nodeType: 'text'
                  }
                ],
                nodeType: 'paragraph'
              }
            ],
            nodeType: 'document'
          }
        },
        teaserDescription: 'oh, shit...Monday came in',
        isContentWithImage: false,
        color: {
          __typename: 'Color',
          ref: 'category-beauty'
        },
        image: {
          __typename: 'Asset',
          title: 'Beauty Category Image',
          url:
            '//images.ctfassets.net/7g2w796onies/3Pa3xwIiNIsKGQB2SE2KYy/9045bf905aba7e22e69054b4419c58a9/category-header-sample.png',
          contentType: 'image/png'
        }
      }
    ]
  }
} as const

const nullGraphQLMock = {
  __typename: 'ContainerOfContentBlocks',
  title: 'Home page Promotion Blocks',
  contentCollection: {
    items: [null, null]
  }
} as const

describe(normalizeContainerOfContentBlocks, () => {
  describe('GraphQL MarketingTeaser blocks', () => {
    it('pulls out the container of content blocks Marketing Teasers', () => {
      expect(normalizeContainerOfContentBlocks(graphQLMock)).toMatchSnapshot()
    })
  })

  describe('GraphQL Promotion blocks', () => {
    it('pulls out the container of content blocks Promotions', () => {
      expect(
        normalizeContainerOfContentBlocks(promotionGraphQLMock)
      ).toMatchSnapshot()
    })
  })
  describe('GraphQL null blocks', () => {
    it('does not error', () => {
      expect(
        normalizeContainerOfContentBlocks(nullGraphQLMock)
      ).toMatchSnapshot()
    })
  })
})
