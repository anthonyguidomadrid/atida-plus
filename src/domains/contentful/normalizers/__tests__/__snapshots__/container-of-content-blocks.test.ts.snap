// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeContainerOfContentBlocks GraphQL MarketingTeaser blocks pulls out the container of content blocks Marketing Teasers 1`] = `
Object {
  "blocks": Array [
    Object {
      "backgroundColor": "category-beauty",
      "contentType": "MarketingTeaser",
      "description": "High quality vitamins for a low price. PT",
      "hasButton": true,
      "isFullWidthImage": false,
      "isSponsoredContent": false,
      "labelColor": "secondary-portland-orange",
      "labelText": "20 % Discount PT",
      "title": "Boost your immunity PT",
      "url": "/boost-your-immunity-pt",
    },
    Object {
      "backgroundColor": "category-nutrition-exercise-and-weight-loss",
      "contentType": "MarketingTeaser",
      "hasButton": false,
      "image": Object {
        "url": "https://images.ctfassets.net/7g2w796onies/6EOfodeKNE2uKYxJTrpbrA/2b730c8a3d7b66e27a4a78863c7a5156/ac6bb2e17841f7dd4ebd7b1fd77f0185.png",
      },
      "isFullWidthImage": false,
      "isSponsoredContent": false,
      "labelColor": "primary-oxford-blue",
      "labelText": "Exclusive offer PT",
      "title": "Save over 50% with Atida Plus PT",
      "url": "/save-over-50-with-atida-plus",
    },
  ],
  "contentType": "ContainerOfContentBlocks",
  "displayTitle": true,
  "hasSmallMargin": false,
  "title": "Home page marketing teasers",
}
`;

exports[`normalizeContainerOfContentBlocks GraphQL Promotion blocks pulls out the container of content blocks Promotions 1`] = `
Object {
  "blocks": Array [
    Object {
      "categories": Array [],
      "color": "category-beauty",
      "contentType": "Promotion",
      "description": "<p>We&#39;re here to help. Ask about our products or get expert medical advice.</p>",
      "id": "Avene has an e with french accent",
      "image": Object {
        "title": "Beauty Category Image",
        "type": "image/png",
        "url": "//images.ctfassets.net/7g2w796onies/3Pa3xwIiNIsKGQB2SE2KYy/9045bf905aba7e22e69054b4419c58a9/category-header-sample.png",
      },
      "isContentWithImage": false,
      "isSponsoredContent": false,
      "labels": Array [
        Object {
          "label": "campaign-promo.black-friday.promo.2021",
          "type": "campaign-promo",
        },
        Object {
          "label": "campaign-promo.black-friday",
          "type": "campaign-promo",
        },
      ],
      "promoInformation": "",
      "teaserDescription": "oh, shit...Monday came in",
      "title": "Avène",
      "url": "",
    },
    Object {
      "categories": Array [],
      "color": "category-beauty",
      "contentType": "Promotion",
      "description": "<p>We&#39;re here to help. Ask about our products or get expert medical advice.</p>",
      "id": "Avene has an e with french accent",
      "image": Object {
        "title": "Beauty Category Image",
        "type": "image/png",
        "url": "//images.ctfassets.net/7g2w796onies/3Pa3xwIiNIsKGQB2SE2KYy/9045bf905aba7e22e69054b4419c58a9/category-header-sample.png",
      },
      "isContentWithImage": false,
      "isSponsoredContent": false,
      "labels": Array [],
      "promoInformation": "",
      "teaserDescription": "oh, shit...Monday came in",
      "title": "Avène",
      "url": "",
    },
  ],
  "contentType": "ContainerOfContentBlocks",
  "displayTitle": false,
  "hasSmallMargin": false,
  "title": "Home page Promotion Blocks",
}
`;

exports[`normalizeContainerOfContentBlocks GraphQL null blocks does not error 1`] = `
Object {
  "blocks": Array [],
  "contentType": "ContainerOfContentBlocks",
  "displayTitle": false,
  "hasSmallMargin": false,
  "title": "Home page Promotion Blocks",
}
`;
